zerver/lib/typed_endpoint.py

```diff
@@ -0,0 +1,514 @@
+        var_name = parameter.request_var_name + "".join(
+            f"[{json.dumps(loc)}]" for loc in error["loc"]
+        )
+        context = {
+            "var_name": var_name,
+            **error.get("ctx", {}),
+        }
+
+        if error["type"] == "json_invalid" and parameter.argument_type_is_body:
+            # argument_type_is_body is usually used by webhooks that do not
+            # require a specific var_name for payload JSON decoding error.
+            # We override it here.
+            error_template = _("Malformed JSON")
+        elif error["type"] in ("unexpected_keyword_argument", "extra_forbidden"):
+            context["argument"] = error["loc"][-1]
+        # This condition matches our StringRequiredConstraint
+        elif error["type"] == "string_too_short" and error["ctx"].get("min_length") == 1:
+            error_template = _("{var_name} cannot be blank")
+
+        assert error_template is not None, MISSING_ERROR_TEMPLATE.format(
+            error_type=error["type"],
+            url=error.get("url", "(documentation unavailable)"),
+            error=json.dumps(error, indent=4),
+        )
+        raise ApiParamValidationError(error_template.format(**context), error["type"])
+
+
+MISSING_ERROR_TEMPLATE = f"""
+    Pydantic validation error of type "{{error_type}}" does not have the
+    corresponding error message template or is not handled explicitly. We expect
+    that every validation error is formatted into a client-facing error message.
+    Consider adding this type to {__package__}.ERROR_TEMPLATES with the appropriate
+    internationalized error message or handle it in {__package__}.{parse_value_for_parameter.__name__}.
+
+    Documentation for "{{error_type}}" can be found at {{url}}.
+
+    Error information:
+{{error}}
+"""
+
+
+def get_func_param(info: ValidationInfo) -> FuncParam[object]:
+    """To access the FuncParam[T] object created when we call
+    parse_value_for_parameter, a custom Pydantic validator can lookup
+    ValidationInfo using this function.
+    See https://docs.pydantic.dev/latest/api/functional_validators for more
+    information on how such validators can be implemented.
+    """
+    assert (
+        info.context is not None and "func_param" in info.context
+    ), "The FuncParam object should always be passed as context when validating"
+    return info.context["func_param"]
+
+
+UNEXPECTEDLY_MISSING_KEYWORD_ONLY_PARAMETERS = """
+Parameters expected to be parsed from the request should be defined as
+keyword-only parameters, but there is no keyword-only parameter found in
+{view_func_name}.
+
+Example usage:
+
+```
+@typed_endpoint
+def view(
+    request: HttpRequest,
+    *,
+    flag_value: Annotated[Json[bool], ApiParamConfig(
+        whence="flag", documentation_status=INTENTIONALLY_UNDOCUMENTED,
+    )]
+) -> HttpResponse:
+    ...
+```
+
+It is likely a programming error unless you only need the ignored parameters check.
```

Comment:
Should this say:

``` quote
This is likely a programming error. See https://peps.python.org/pep-3102/ for details on how
to correctly declare your parameters as keyword-only parameters.
Endpoints that do not accept parameters should use @typed_endpoint_without_parameters.
```

Reply:
Yeah it's definitely nice to have the link there.